ksql-engine/src/main/java/io/confluent/ksql/KsqlExecutionContext.java

```diff
@@ -46,6 +47,11 @@
    */
   MetaStore getMetaStore();
 
+  /**
+   * @return the service context used for this execution context
+   */
+  ServiceContext getServiceContext();
```

Comment:
I had the same instinct as you, in fact that was the way I originally had it. The problem was that not just any sandboxed service context needs to be used - it has to be exactly the same sandboxed context that the sandboxed engine uses. Otherwise it can't validate a sequence of statements that create topics, for example:

```sql
0: CREATE STREAM Y AS SELECT * FROM X; 
1: CREATE STREAM Z AS SELECT * FROM Y;
```

When command `0` runs, it updates the service context with a new topic `Y` with `p` partitions and `r` replicas. When command `1` runs, it expects to be able to read from the topic for `Y`, but that topic doesn't exist.

So, unfortunately, the coupling is an artifact of where the sandbox is created (deep inside the engine). 

To do what you suggest, we can pipe in `ServiceContext` through the `ksqlEngine#createSandbox`, into the constructor for `SandboxedExecutionContext` and then into `EngineContext#createSandbox` from every place we call `createSandbox`.

This isn't so bad, but it's tricky to get right (to make sure that we always use the same execution context as the engine we are using). As soon as I thought about that, I decided that it was appropriate to expose the service context because (as it stands) the coupling is actually an attribute of the engine, not just an implementation detail.

If you strongly disagree, I can do one of two things:
* Expose an `Optional<ServiceContext> getSandboxContext` which will only return if it is a sandbox (ew 🤢 )
* Do the refactor that I outlined above but risk loose coupling of the context and the engine

Reply:
Hummm... let's see how it looks.

Reply:
The impersonation work that Sergio is doing will mean that each request to the engine will need to come with its own `ServiceContext` that runs things in the context of the initiating user. 

With that in mind, it probably makes sense to not expose the service context from the engine, but instead pass it each time.